plugin-atlas/src/main/java/org/apache/ranger/authorization/atlas/authorizer/RangerAtlasAuthorizerUtil.java

```diff
@@ -78,57 +79,71 @@ static void collectAccessors(RangerAccessResult result, AtlasAccessorResponse re
 
             result.getMatchedItems().forEach(x -> {
-                response.getUsers().addAll(x.getUsers());
-                response.getRoles().addAll(x.getRoles());
-                response.getGroups().addAll(x.getGroups());
+                collectSubjects(response, x);
             });
         }
     }
 
+    static private void collectSubjects(AtlasAccessorResponse response, RangerPolicyItemEvaluator evaluator) {
+
+        RangerPolicy.RangerPolicyItem policyItem = evaluator.getPolicyItem();
+
+        if (evaluator.getPolicyItemType() == RangerPolicyItemEvaluator.POLICY_ITEM_TYPE_ALLOW) {
+            response.getUsers().addAll(policyItem.getUsers());
+            response.getRoles().addAll(policyItem.getRoles());
+            response.getGroups().addAll(policyItem.getGroups());
+
+        } else if (evaluator.getPolicyItemType() == RangerPolicyItemEvaluator.POLICY_ITEM_TYPE_DENY) {
+            response.getDenyUsers().addAll(policyItem.getUsers());
+            response.getDenyRoles().addAll(policyItem.getRoles());
+            response.getDenyGroups().addAll(policyItem.getGroups());
+        }
+    }
+
     static void collectAccessors(RangerAccessResult resultEnd1, RangerAccessResult resultEnd2, AtlasAccessorResponse accessorResponse) {
 
         if (resultEnd2 == null || CollectionUtils.isEmpty(resultEnd2.getMatchedItems()))  {
             return;
         }
 
-        final List<String> usersEnd1 = new ArrayList<>();
-        final List<String> rolesEnd1 = new ArrayList<>();
-        final List<String> groupsEnd1 = new ArrayList<>();
-
-        final List<String> usersEnd2 = new ArrayList<>();
-        final List<String> rolesEnd2 = new ArrayList<>();
-        final List<String> groupsEnd2 = new ArrayList<>();
+        final AtlasAccessorResponse accessorsEnd1 = new AtlasAccessorResponse();
+        final AtlasAccessorResponse accessorsEnd2 = new AtlasAccessorResponse();
 
         // Collect lists of accessors for both results
         resultEnd1.getMatchedItems().forEach(x -> {
-            usersEnd1.addAll(x.getUsers());
-            rolesEnd1.addAll(x.getRoles());
-            groupsEnd1.addAll(x.getGroups());
+            collectSubjects(accessorsEnd1, x);
         });
 
         resultEnd2.getMatchedItems().forEach(x -> {
-            usersEnd2.addAll(x.getUsers());
-            rolesEnd2.addAll(x.getRoles());
-            groupsEnd2.addAll(x.getGroups());
+            collectSubjects(accessorsEnd2, x);
         });
 
         // Retain only common accessors
-        usersEnd1.retainAll(usersEnd2);
-        rolesEnd1.retainAll(rolesEnd2);
-        groupsEnd1.retainAll(groupsEnd2);
+        accessorsEnd1.getUsers().retainAll(accessorsEnd2.getUsers());
+        accessorsEnd1.getRoles().retainAll(accessorsEnd2.getRoles());
+        accessorsEnd1.getGroups().retainAll(accessorsEnd2.getGroups());
+
+        accessorsEnd1.getDenyUsers().retainAll(accessorsEnd2.getDenyUsers());
```

Comment:
In case of deny - Shouldn't we return a union of end1 and end2 instead of an intersection?

Reply:
This is specific for tag-based policies(TBP) for relationship actions.

How does TBP allow access to relationship action?
-> There must be TBP/policies which allow both ends to the same user
There might be different users per allowed TBP when a couple of tag-based policies allow a user to perform an action on a relationship

So it is always intersection not union to collect common users from both TBP allowed items

The same applies to in case of Deny policy items.
We are returning users which are explicitly denied an action & it also needs an intersection of subjects